### PR TITLE
Add provider access scoping to /transactions client

### PR DESCRIPTION
**Pull request by cjlynch12**
## 📚 Purpose
Adds validation for `transactions:read:provider` claim to `/transactions`.
Will address the other mds-transaction-api routes separately. For now they will continue to only accept the `transactions:read` claim.

## 📦 Impacts:
mds-transaction-api

### CODE REVIEW

File: packages/mds-transaction-api/@types/index.ts [hunk 1]
```diff
@@ -23,7 +23,11 @@ export const [TRANSACTION_API_DEFAULT_VERSION] = TRANSACTION_API_SUPPORTED_VERSI
 // Allow adding type definitions for Express Request objects
 export type TransactionApiRequest<B = {}> = ApiRequest<B>
 
-export type TransactionApiAccessTokenScopes = 'transactions:read' | 'transactions:read:claim' | 'transactions:write'
+export type TransactionApiAccessTokenScopes =
+  | 'transactions:read'
+  | 'transactions:read:provider'
+  | 'transactions:read:claim'
```
Comment (drtyh2o) [inline packages/mds-transaction-api/@types/index.ts:29]: I think `transactions:read:claim` isn't being used and was probably meant to indicate that transactions should be filtered by claims, such as `provider_id` and perhaps one day `agency` or `jurisdiction`, however, I think you should proceed with adding `transactions:read:provider`, but remove `transactions:read:claim`.

File: packages/mds-transaction-api/handlers/get-transactions.ts [hunk 2]
```diff
@@ -118,10 +118,23 @@ export const GetTransactionsHandler = async (
...
+    const provider_id = scopes.includes('transactions:read') ? quereied_provider_id : res.locals.claims?.provider_id
+
+    if (provider_id === null) {
+      /* This should never happen -- a client with just the transactions:read:provider scope
+       * should *always* have a provider_id claim in their token.
+       */
+
+      return res.status(500).send({ error: 'internal_server_error' })
+    }
+
```
Comment (drtyh2o) [inline packages/mds-transaction-api/handlers/get-transactions.ts:137]: In at least one other instance we have used the `non-null assertion` operator in this scenario when other middleware has previously asserted something exists.

**Review by drtyh2o — APPROVED** [review]

File: packages/mds-transaction-api/handlers/get-transactions.ts (hunk 2 above)
Comment (cjlynch12) [reply packages/mds-transaction-api/handlers/get-transactions.ts:137]: Yeah saw both this and the `non-null assertion` being used, so took a gamble. Will happily change to `non-null assertion`.

File: packages/mds-transaction-api/handlers/get-transactions.ts [hunk 3]
```diff
@@ -118,10 +118,23 @@ export const GetTransactionsHandler = async (
   next: express.NextFunction
 ) => {
   try {
+    const { scopes } = res.locals
+
     const order = getOrderOption(req)
-    const { provider_id, before, after } = parseRequest(req)
+    const { provider_id: quereied_provider_id, before, after } = parseRequest(req)
```
Comment (avatarneil) [inline packages/mds-transaction-api/handlers/get-transactions.ts:124]: `quereied_provider_id` -> `queried_provider_id`

File: packages/mds-transaction-api/handlers/get-transactions.ts [hunk 4]
```diff
@@ -118,10 +118,23 @@ export const GetTransactionsHandler = async (
...
       .query('provider_id', 'before', 'after')
+
+    const provider_id = scopes.includes('transactions:read') ? quereied_provider_id : res.locals.claims?.provider_id
+
+    if (provider_id === null) {
+      /* This should never happen -- a client with just the transactions:read:provider scope
+       * should *always* have a provider_id claim in their token.
+       */
+
+      return res.status(500).send({ error: 'internal_server_error' })
```
Comment (avatarneil) [inline packages/mds-transaction-api/handlers/get-transactions.ts:135]: I'd recommend leaning on the global error handling middleware here, and throwing an error. Also, definitely want to log out exactly what's going on, cause if this _does_ happen we wanna know what's up.

**Review by avatarneil — CHANGES_REQUESTED** [review]
couple changes I'd like to see before this hits mainstream

File: packages/mds-transaction-api/handlers/get-transactions.ts (hunk 4 above)
Comment (cjlynch12) [reply packages/mds-transaction-api/handlers/get-transactions.ts]: Replaced in favor of adding `claims?.provider_id` validation to the middleware using `isUUID()`.

**Review by nessur — APPROVED** [review]

**Review by avatarneil — APPROVED** [review]

**Pull request merged**